external/java/src/main/java/ch/epfl/dedis/skipchain/SkipchainRPC.java

```diff
@@ -74,39 +84,114 @@ public SkipBlock getSkipblock(SkipblockId id) throws CothorityCommunicationExcep
 
         try {
             SkipchainProto.SkipBlock sb = SkipchainProto.SkipBlock.parseFrom(msg);
-            //TODO: add verification that the skipblock is valid by hashing and comparing to the id
+            SkipBlock ret = new SkipBlock(sb);
+
+            // simple verification (we do not check the links, just the signature)
+            if (!ret.verifyForwardSignatures()) {
+                throw new CothorityCommunicationException("invalid forward signatures");
+            }
 
             logger.debug("Got the following skipblock: {}", sb);
             logger.info("Successfully read skipblock");
 
-            return new SkipBlock(sb);
+            return ret;
         } catch (InvalidProtocolBufferException e) {
             throw new CothorityCommunicationException(e);
         }
     }
 
-    public SkipBlock getLatestSkipblock() throws CothorityCommunicationException {
-        SkipchainProto.GetUpdateChain request =
-                SkipchainProto.GetUpdateChain.newBuilder().setLatestID(ByteString.copyFrom(scID.getId())).build();
+    /**
+     * Returns the latest block and verify that the links are correct. The verification is performed from the genesis block.
+     *
+     * @return the latest skipblock
+     * @throws CothorityCommunicationException if something goes wrong with communication
+     * @throws CothorityCryptoException if the verification goes wrong
+     */
+    public SkipBlock getLatestSkipblock() throws CothorityCommunicationException, CothorityCryptoException {
+        return this.getLatestSkipblock(this.scID);
+    }
 
-        ByteString msg = roster.sendMessage("Skipchain/GetUpdateChain",
-                request);
+    /**
```

Comment:
I'm wondering why this code has to be written to work only with the level-0 links. It we can find a path back to the trustedLatest via higher level links, that's better right?

I think I understand that this is the simplest thing that works, but it feels like we are losing the benefits of skipchain right here....

Reply:
I'm a bit confused by this comment. I don't think this code only works for level-0 links. The request doesn't specify the maximum height. So what is returned to us are all the blocks that are linked by the highest forward link. Then we only check the hash on those highest links (`https://github.com/dedis/cothority/pull/1649/files#diff-e1b7d9d6ed776d2c287512dce30b3a63L105`)